gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/service/token/impl/TokenServiceImpl.java

```diff
@@ -126,9 +126,24 @@ public Maybe<Token> getRefreshToken(String refreshToken, Client client) {
     }
 
     @Override
-    public Single<Token> introspect(String token) {
-        return introspectionTokenService.introspect(token, false)
-                .map(this::convertAccessToken);
+    public Maybe<Token> introspect(String token) {
+        return introspectionTokenFacade.introspectAccessToken(token)
+                .map(this::convertAccessToken)
+                .switchIfEmpty(introspectionTokenFacade.introspectRefreshToken(token).map(this::convertRefreshToken));
+    }
+
+    @Override
+    public Maybe<Token> introspect(String token, TokenTypeHint hint) {
+        switch (hint) {
+            case REFRESH_TOKEN:
+                return introspectionTokenFacade.introspectRefreshToken(token)
+                        .map(this::convertRefreshToken)
+                        .switchIfEmpty(introspectionTokenFacade.introspectAccessToken(token).map(this::convertAccessToken));
+            case ID_TOKEN:
```

Comment:
**thought**: is it useful to process ID_TOKEN ? IMO it can be the default behaviour by returning Maybe.empty() because introspect is not about id_token and anyway this token is currently not persisted in the DB so we can save access to the storage layer.

Reply:
right. I fixed that